### PR TITLE
Table writer context

### PR DESCRIPTION
**Pull request by calispac**
I suppose this solves an issue in #669.

I created abstract methods in the TableWrite and TableReader:

- close()
- open()

the __enter__() and __exit__() methods are handled by the parent class.

### CODE REVIEW

**Comment by calispac** [comment]
I just need to run some tests. Not sure it works as supposed yet.

**Comment by kosack** [comment]
could you also add a small unit test to  test this functionality in `ctapipe/io/tests/test_hdf5.py`?

**Comment by calispac** [comment]
> could you also add a small unit test to test this functionality in ctapipe/io/tests/test_hdf5.py?
Sure!

**Comment by MaxNoe** [comment]
I would completely remove the `__del__` method.

**Comment by calispac** [comment]
I have added some more. I would like to test the fact that the __exit__() method really closes the file but I haven't found a way to do that...
Anyway it should be in the TableWriter and TableReader.

**Comment by MaxNoe** [comment]
> I would like to test the fact that the exit()

The reference to the object still is there after the `with` block, so you can check if `writer._file.is_open` or whatever the underlying file provides

File: ctapipe/io/hdftableio.py [hunk 1]
```diff
@@ -385,9 +428,18 @@ def __init__(self, filename):
...
         super().__init__()
         self._tables = {}
-        self._h5file = tables.open_file(filename)
+        self._h5file = self.open(filename)
         pass
 
+    def open(self, filename):
+
+        self._h5file = tables.open_file(filename)
+        return self._h5file
```
Comment (MaxNoe) [inline ctapipe/io/hdftableio.py]: why the return?

Comment (calispac) [reply ctapipe/io/hdftableio.py]: yes I think we can remove it

Comment (calispac) [reply ctapipe/io/hdftableio.py]: hum actully I don't like much that class attributes are definded outside of the __init__()

so I will make open() return the table without assigning it to the class attribute

Comment (MaxNoe) [reply ctapipe/io/hdftableio.py]: I don't think this makes sense. 

I think it's perfectly alright for a method called `open` to set internal state of a class

**Comment by MaxNoe** [comment]
Sorry, wrong button. I should'n even be allowed to do this as a non-maintainer

File: ctapipe/io/hdftableio.py (hunk 1 above)
Comment (calispac) [reply ctapipe/io/hdftableio.py]: ok!

**Comment by calispac** [comment]
I believe it is ready. Just tell me if there is anything left.
Thanks for the fast review

**Comment by MaxNoe** [comment]
Maybe a question for another PR, but does tables support appending? So could one add the mode to the args if `writer`? To make appending possible?

**Comment by calispac** [comment]
> Maybe` a question for another PR, but does tables support appending? So could one add mode to the args of writer? To make appending possible?

I believe that looking at the current structure of Writer/Reader one should then have an "Appender". Therefore maybe this can be revised having only a HDF5Parser class which handles all the `mode`?

File: ctapipe/io/hdftableio.py [hunk 2]
```diff
@@ -91,6 +99,16 @@ def write(self, table_name, containers):
         """
         pass
 
+    @abstractmethod
+    def open(self, filename):
```
Comment (kosack) [inline ctapipe/io/hdftableio.py]: need to add  **kwargs here, since the parameters of the overridden methods must match the base class

Comment (kosack) [reply ctapipe/io/hdftableio.py]: that's an old mistake, but might as well finish it now

Comment (MaxNoe) [reply ctapipe/io/hdftableio.py]: I this is not true. Python (at least not abc) does not check the method's signature:

```
from abc import ABCMeta, abstractmethod


class A(metaclass=ABCMeta):

    @abstractmethod
    def test(self):
        pass


class B(A):

    def test(self, name):
        print('Hello', name + '!')


b = B()
b.test('World')
```

**Pull request merged**